example/mapboxExampleCamera.js

```diff
@@ -0,0 +1,284 @@
+
+				_this.setupCamera();
+
+			} );
+
+		this.setupCamera();
+
+	}
+
+	setupCamera() {
+
+		const t = this.map.transform;
+		// if aspect is not reset raycast will fail on map resize
+		this.camera.aspect = t.width / t.height;
+		this.halfFov = t._fov / 2;
+		this.cameraToCenterDistance = ( 0.5 / Math.tan( this.halfFov ) ) * t.height;
+		const maxPitch = ( t._maxPitch * Math.PI ) / 180;
+		this.acuteAngle = Math.PI / 2 - maxPitch;
+		this.updateCamera();
+
+	}
+
+	updateCamera( ev ) {
+
+		if ( ! this.camera ) {
+
+			console.log( 'nocamera' );
+			return;
+
+		}
+
+		const t = this.map.transform;
+
+		this.camera.aspect = t.width / t.height;
+		let farZ = 0;
+		let furthestDistance = 0;
+		this.halfFov = t._fov / 2;
+		const groundAngle = Math.PI / 2 + t._pitch;
+		// pitch seems to influence heavily the depth calculation and cannot be more than 60 = PI/3 < v1 and 85 > v2
+		const pitchAngle = Math.cos( Math.PI / 2 - t._pitch );
+		this.cameraToCenterDistance = ( 0.5 / Math.tan( this.halfFov ) ) * t.height;
+
+		let pixelsPerMeter = 1;
+		const worldSize = this.worldSize();
+
+		pixelsPerMeter = this.mercatorZfromAltitude( 1, t.center.lat ) * worldSize;
+		const fovAboveCenter = t._fov * ( 0.5 + t.centerOffset.y / t.height );
+
+		// Adjust distance to MSL by the minimum possible elevation visible on screen,
+		// this way the far plane is pushed further in the case of negative elevation.
+		const minElevationInPixels = t.elevation
+			? t.elevation.getMinElevationBelowMSL() * pixelsPerMeter
+			: 0;
+
+		const cameraToSeaLevelDistance = ( t._camera.position[ 2 ] * worldSize - minElevationInPixels ) / Math.cos( t._pitch );
+
+		const topHalfSurfaceDistance = ( Math.sin( fovAboveCenter ) * cameraToSeaLevelDistance ) / Math.sin(
+			utils.clamp(
+				Math.PI - groundAngle - fovAboveCenter,
+				0.01,
+				Math.PI - 0.01
+			)
+		);
+
+		// Calculate z distance of the farthest fragment that should be rendered.
+		furthestDistance = pitchAngle * topHalfSurfaceDistance + cameraToSeaLevelDistance;
+		// Add a bit extra to avoid precision problems when a fragment's distance is exactly `furthestDistance`
+		const horizonDistance = cameraToSeaLevelDistance * ( 1 / t._horizonShift );
+		farZ = Math.min( furthestDistance * 1.01, horizonDistance );
+
+		this.cameraTranslateZ = new THREE.Matrix4().makeTranslation(
+			0,
+			0,
+			this.cameraToCenterDistance
+		);
+		// someday @ansis set further near plane to fix precision for deckgl,so we should fix it to use mapbox-gl v1.3+ correctly
+		// https://github.com/mapbox/mapbox-gl-js/commit/5cf6e5f523611bea61dae155db19a7cb19eb825c#diff-5dddfe9d7b5b4413ee54284bc1f7966d
+		const nz = t.height / 50; // min near z as coded by @ansis
+		const nearZ = Math.max( nz * pitchAngle, nz ); // on changes in the pitch nz could be too low
+
+		const h = t.height;
+		const w = t.width;
+		this.camera.projectionMatrix = utils.makePerspectiveMatrix(
+			t._fov,
+			w / h,
+			nearZ,
+			farZ
+		);
```

Comment:
There shouldn't be a need to manually construct a perspective matrix. These are all values that can be set on the camera itself and "updateProjectionMatrix" can be called to correctly update the projection matrix and ensure the inverse projection matrix are set

Reply:
Unlike the Mapbox GL JS camera, separate camera translation and rotation out into its world matrix If this is applied directly to the projection matrix, it will work OK but break raycasting

Reply:
I'm not following. I don't think we should include position or rotation values in the projection matrix. With the values you have here you can set the three.js camera parameters like so:

```js
camera.fov = t._fov;
camera.aspect = w / h;
camera.near = nearZ;
camera.far = farZ;
camera.updateProjectionMatrix();
```

Reply:
We need to use the same projectionMatrix as mapboxgl. I modified this place and used the method this.camera.projectionMatrix.elements = t._camera.getCameraToClipPerspective provided by mapboxgl. I found that the calculation methods of projectionMatrix of mapboxgl and threejs are somewhat different. Mapboxgl uses the gl-matrix library.

Reply:
I've taken a look at the gl-matrix code [here](https://glmatrix.net/docs/mat4.js.html#line1474) and the three.js projection matrix code [here]() and I don't see any difference in calculations beyond superficial organization. Can you explain what the differences you saw are?